In [57]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [58]:
import warnings
warnings.filterwarnings('ignore')

In [59]:
df = pd.read_csv('https://raw.githubusercontent.com/dsrscientist/Data-Science-ML-Capstone-Projects/master/baseball.csv')
df.head()

W    R    AB     H   2B  3B   HR   BB    SO   SB   RA   ER   ERA  CG  SHO  \
0  95  724  5575  1497  300  42  139  383   973  104  641  601  3.73   2    8   
1  83  696  5467  1349  277  44  156  439  1264   70  700  653  4.07   2   12   
2  81  669  5439  1395  303  29  141  533  1157   86  640  584  3.67  11   10   
3  76  622  5533  1381  260  27  136  404  1231   68  701  643  3.98   7    9   
4  74  689  5605  1515  289  49  151  455  1259   83  803  746  4.64   7   12   

   SV    E  
0  56   88  
1  45   86  
2  38   79  
3  37  101  
4  35   86

In [60]:
df.shape

(30, 17)

In [61]:
df.isna().sum()

W      0
R      0
AB     0
H      0
2B     0
3B     0
HR     0
BB     0
SO     0
SB     0
RA     0
ER     0
ERA    0
CG     0
SHO    0
SV     0
E      0
dtype: int64

In [62]:
df.describe()

W           R           AB            H          2B  \
count   30.000000   30.000000    30.000000    30.000000   30.000000   
mean    80.966667  688.233333  5516.266667  1403.533333  274.733333   
std     10.453455   58.761754    70.467372    57.140923   18.095405   
min     63.000000  573.000000  5385.000000  1324.000000  236.000000   
25%     74.000000  651.250000  5464.000000  1363.000000  262.250000   
50%     81.000000  689.000000  5510.000000  1382.500000  275.500000   
75%     87.750000  718.250000  5570.000000  1451.500000  288.750000   
max    100.000000  891.000000  5649.000000  1515.000000  308.000000   

              3B          HR          BB          SO          SB          RA  \
count  30.000000   30.000000   30.000000    30.00000   30.000000   30.000000   
mean   31.300000  163.633333  469.100000  1248.20000   83.500000  688.233333   
std    10.452355   31.823309   57.053725   103.75947   22.815225   72.108005   
min    13.000000  100.000000  375.000000   973.00000   44.000000  525.000000   
25%    23.000000  140.250000  428.250000  1157.50000   69.000000  636.250000   
50%    31.000000  158.500000  473.000000  1261.50000   83.500000  695.500000   
75%    39.000000  177.000000  501.250000  1311.50000   96.500000  732.500000   
max    49.000000  232.000000  570.000000  1518.00000  134.000000  844.000000   

               ER        ERA         CG        SHO         SV           E  
count   30.000000  30.000000  30.000000  30.000000  30.000000   30.000000  
mean   635.833333   3.956333   3.466667  11.300000  43.066667   94.333333  
std     70.140786   0.454089   2.763473   4.120177   7.869335   13.958889  
min    478.000000   2.940000   0.000000   4.000000  28.000000   75.000000  
25%    587.250000   3.682500   1.000000   9.000000  37.250000   86.000000  
50%    644.500000   4.025000   3.000000  12.000000  42.000000   91.000000  
75%    679.250000   4.220000   5.750000  13.000000  46.750000   96.750000  
max    799.000000   5.040000  11.000000  21.000000  62.000000  126.000000

In [63]:
from sklearn.preprocessing import PowerTransformer

In [64]:
features = ['W','R','AB','H','2B','3B','HR','BB','SO','SB','RA','ER','ERA','CG','SHO','SV','E']

In [65]:
scaler = PowerTransformer(method='yeo-johnson')

In [66]:
df[features] = scaler.fit_transform(df[features].values)

In [67]:
df[features]

W    R   AB    H        2B        3B        HR        BB        SO  \
0   1.346947  0.0  0.0  0.0  1.477685  1.033103 -0.764626 -1.610915 -2.569896   
1   0.218658  0.0  0.0  0.0  0.084269  1.203320 -0.158581 -0.502749  0.134913   
2   0.025271  0.0  0.0  0.0  1.670782 -0.160270 -0.689676  1.128294 -0.901254   
3  -0.465574  0.0  0.0  0.0 -0.848600 -0.360521 -0.879049 -1.181207 -0.189628   
4  -0.665017  0.0  0.0  0.0  0.792183  1.616444 -0.329850 -0.206771  0.085459   
5   1.162486  0.0  0.0  0.0  1.998506 -1.461062  1.937860  1.708067 -0.957960   
6   0.600668  0.0  0.0  0.0 -0.198558 -1.224911  1.460099  1.461121 -0.228668   
7   0.025271  0.0  0.0  0.0 -1.556454 -1.110302  1.583551 -0.904450  0.807174   
8  -0.072041  0.0  0.0  0.0  0.141687  0.130611  0.200058 -0.559191  0.594561   
9  -0.267936  0.0  0.0  0.0  1.099405  0.225256  0.007404  0.204599 -0.986257   
10  0.695214  0.0  0.0  0.0  0.199390  0.130611  0.355503  0.634349 -0.170084   
11  0.505745  0.0  0.0  0.0  0.141687 -0.462732  1.891937  0.344014  1.434475   
12  0.410440  0.0  0.0  0.0 -1.701159 -0.997802  0.476708 -0.578074 -0.967396   
13 -0.465574  0.0  0.0  0.0 -0.743057 -0.887277  1.098636  0.204599  0.858050   
14 -1.274850  0.0  0.0  0.0  0.084269  1.370614 -0.506753  0.151842 -1.257838   
15  1.802315  0.0  0.0  0.0  0.731608  0.771872 -0.840636  0.684777  0.164634   
16  1.621136  0.0  0.0  0.0  0.975645 -0.360521 -0.727020 -0.097899  0.715843   
17  1.530067  0.0  0.0  0.0 -0.198558 -0.062101  0.324770  1.662186  2.774844   
18 -1.274850  0.0  0.0  0.0 -0.086278  0.318822 -0.542843 -1.022149  0.483883   
19 -1.691747  0.0  0.0  0.0 -1.004830 -0.360521  0.200058  0.515754  0.045967   
20  0.883198  0.0  0.0  0.0  1.161721 -1.461062  0.506589  0.378584  0.393683   
21  0.218658  0.0  0.0  0.0 -0.582652 -1.966661  0.506589  1.224431  0.939655   
22 -0.967709  0.0  0.0  0.0 -2.029345  0.859772 -1.534280 -1.779520 -0.967396   
23 -1.378263  0.0  0.0  0.0 -1.309831 -1.341775 -2.485131  0.081088 -1.369174   
24 -1.797360  0.0  0.0  0.0 -0.198558  0.593462 -1.115481 -1.527654  0.234121   
25  1.069741  0.0  0.0  0.0 -0.689868 -0.462732  0.796626  1.600712  0.075580   
26  0.314746  0.0  0.0  0.0  0.731608  0.771872 -0.879049 -0.170355 -0.882318   
27 -0.169774  0.0  0.0  0.0  0.792183  1.535156 -0.226436  0.413043  0.614735   
28 -0.665017  0.0  0.0  0.0 -0.848600  0.502887 -0.435292 -0.749579  0.766543   
29 -1.274850  0.0  0.0  0.0 -0.086278  1.616444  0.768314 -1.506945  0.323750   

          SB        RA        ER       ERA        CG       SHO        SV  \
0   0.939708 -0.660331 -0.497425 -0.500893 -0.359844 -0.789579  1.546378   
1  -0.539693  0.175465  0.257974  0.261806 -0.359844  0.251775  0.360444   
2   0.211956 -0.674621 -0.746906 -0.636459  1.993513 -0.245393 -0.607594   
3  -0.642564  0.189512  0.113575  0.060804  1.220103 -0.510759 -0.764453   
4   0.079629  1.603228  1.582654  1.520959  1.220103  0.251775 -1.095010   
5   0.298238 -0.247757 -0.380464 -0.343110  1.220103 -0.245393 -1.269438   
6  -0.910154  0.147360  0.243552  0.172549  0.064630 -2.099326  0.718082   
7  -2.102897  0.077029  0.156937  0.217193 -1.729889 -0.245393  0.104606   
8   0.255285 -0.646046 -0.453532 -0.478328 -0.909181  0.251775  1.905480   
9  -0.489090  0.914815  0.845910  0.794823  0.064630 -0.245393 -0.309109   
10  0.825451  0.637024  0.645876  0.639800  0.716445 -0.510759  0.360444   
11  1.543609 -0.990086 -0.923801 -0.863078  0.716445  0.486160 -0.455910   
12 -1.560880 -0.176972 -0.074751 -0.028733 -0.359844  0.251775  0.482968   
13 -0.590846  0.539455  0.602915  0.462190  0.981874  0.251775  0.360444   
14 -0.149220  0.581292  0.416351  0.417713  0.716445  0.931345 -2.473461   
15 -0.590846 -2.348646 -2.334150 -2.311585 -0.909181  0.931345  2.073674   
16  0.708582 -1.307707 -1.518396 -1.686213 -1.729889  0.486160  1.354219   
17  0.588959 -1.134186 -1.309398 -1.341812  0.981874  2.122859  0.718082   
18  0.124135  0.692697  0.674497  0.728428 -0.909181 -1

In [68]:
X = df[features].drop('W',axis=1)

In [69]:
y =df[features].W

In [70]:
X

R   AB    H        2B        3B        HR        BB        SO        SB  \
0   0.0  0.0  0.0  1.477685  1.033103 -0.764626 -1.610915 -2.569896  0.939708   
1   0.0  0.0  0.0  0.084269  1.203320 -0.158581 -0.502749  0.134913 -0.539693   
2   0.0  0.0  0.0  1.670782 -0.160270 -0.689676  1.128294 -0.901254  0.211956   
3   0.0  0.0  0.0 -0.848600 -0.360521 -0.879049 -1.181207 -0.189628 -0.642564   
4   0.0  0.0  0.0  0.792183  1.616444 -0.329850 -0.206771  0.085459  0.079629   
5   0.0  0.0  0.0  1.998506 -1.461062  1.937860  1.708067 -0.957960  0.298238   
6   0.0  0.0  0.0 -0.198558 -1.224911  1.460099  1.461121 -0.228668 -0.910154   
7   0.0  0.0  0.0 -1.556454 -1.110302  1.583551 -0.904450  0.807174 -2.102897   
8   0.0  0.0  0.0  0.141687  0.130611  0.200058 -0.559191  0.594561  0.255285   
9   0.0  0.0  0.0  1.099405  0.225256  0.007404  0.204599 -0.986257 -0.489090   
10  0.0  0.0  0.0  0.199390  0.130611  0.355503  0.634349 -0.170084  0.825451   
11  0.0  0.0  0.0  0.141687 -0.462732  1.891937  0.344014  1.434475  1.543609   
12  0.0  0.0  0.0 -1.701159 -0.997802  0.476708 -0.578074 -0.967396 -1.560880   
13  0.0  0.0  0.0 -0.743057 -0.887277  1.098636  0.204599  0.858050 -0.590846   
14  0.0  0.0  0.0  0.084269  1.370614 -0.506753  0.151842 -1.257838 -0.149220   
15  0.0  0.0  0.0  0.731608  0.771872 -0.840636  0.684777  0.164634 -0.590846   
16  0.0  0.0  0.0  0.975645 -0.360521 -0.727020 -0.097899  0.715843  0.708582   
17  0.0  0.0  0.0 -0.198558 -0.062101  0.324770  1.662186  2.774844  0.588959   
18  0.0  0.0  0.0 -0.086278  0.318822 -0.542843 -1.022149  0.483883  0.124135   
19  0.0  0.0  0.0 -1.004830 -0.360521  0.200058  0.515754  0.045967  1.963116   
20  0.0  0.0  0.0  1.161721 -1.461062  0.506589  0.378584  0.393683 -1.625038   
21  0.0  0.0  0.0 -0.582652 -1.966661  0.506589  1.224431  0.939655 -1.253362   
22  0.0  0.0  0.0 -2.029345  0.859772 -1.534280 -1.779520 -0.967396  1.232586   
23  0.0  0.0  0.0 -1.309831 -1.341775 -2.485131  0.081088 -1.369174 -0.590846   
24  0.0  0.0  0.0 -0.198558  0.593462 -1.115481 -1.527654  0.234121  0.298238   
25  0.0  0.0  0.0 -0.689868 -0.462732  0.796626  1.600712  0.075580 -1.136048   
26  0.0  0.0  0.0  0.731608  0.771872 -0.879049 -0.170355 -0.882318  0.507605   
27  0.0  0.0  0.0  0.792183  1.535156 -0.226436  0.413043  0.614735  1.900657   
28  0.0  0.0  0.0 -0.848600  0.502887 -0.435292 -0.749579  0.766543  0.034713   
29  0.0  0.0  0.0 -0.086278  1.616444  0.768314 -1.506945  0.323750  0.669021   

          RA        ER       ERA        CG       SHO        SV         E  
0  -0.660331 -0.497425 -0.500893 -0.359844 -0.789579  1.546378 -0.334953  
1   0.175465  0.257974  0.261806 -0.359844  0.251775  0.360444 -0.532211  
2  -0.674621 -0.746906 -0.636459  1.993513 -0.245393 -0.607594 -1.342331  
3   0.189512  0.113575  0.060804  1.220103 -0.510759 -0.764453  0.678524  
4   1.603228  1.582654  1.520959  1.220103  0.251775 -1.095010 -0.532211  
5  -0.247757 -0.380464 -0.343110  1.220103 -0.245393 -1.269438 -0.334953  
6   0.147360  0.243552  0.172549  0.064630 -2.099326  0.718082  0.104539  
7   0.077029  0.156937  0.217193 -1.729889 -0.245393  0.104606 -1.614550  
8  -0.646046 -0.453532 -0.478328 -0.909181  0.251775  1.905480  0.261504  
9   0.914815  0.845910  0.794823  0.064630 -0.245393 -0.309109  0.409004  
10  0.637024  0.645876  0.639800  0.716445 -0.510759  0.360444  1.577725  
11 -0.990086 -0.923801 -0.863078  0.716445  0.486160 -0.455910 -0.636014  
12 -0.176972 -0.074751 -0.028733 -0.359844  0.251775  0.482968  0.104539  
13  0.539455  0.602915  0.462190  0.981874  0.251775  0.360444  0.184254  
14  0.581292  0.416351  0.417713  0.716445  0.931345 -2.473461  1.829127  
15 -2.348646 -2.334150 -2.311585 -0.909181  0.931345  2.073674  0.336388  
16 -1.307707 -1.518396 -1.686213 -1.729889  0.486160  1.354219  1.690704  
17 -1.134186 -1.309398 -1.341812  0.981874  2.122859  0.718082  1.231063  
18  0.692697  0.674497  0.728428 -0.909181 -1.084158 -0.309109  1.455

In [71]:
from sklearn.preprocessing import StandardScaler

In [72]:
scaler = StandardScaler()

In [73]:
X_scaler = scaler.fit_transform(X)

In [74]:
X_train,X_test,y_train,y_test = train_test_split(X_scaler,y,test_size=0.25,random_state=355)

In [75]:
from sklearn.linear_model import LinearRegression

In [76]:
linreg = LinearRegression()
linreg.fit(X_train,y_train)

LinearRegression()

In [77]:
y1_pred = linreg.predict(X_test)

In [78]:
from sklearn.metrics import mean_absolute_error,mean_squared_error
print('MAE for testing set: {}'.format(mean_absolute_error(y1_pred,y_test)))
print('MSE for testing set: {}'.format(mean_squared_error(y1_pred,y_test)))
print('RMSE for testing set: {}'.format(np.sqrt(mean_squared_error(y1_pred,y_test))))

MAE for testing set: 0.5715350770042953
MSE for testing set: 0.43088556545062007
RMSE for testing set: 0.6564187424583641


In [79]:
from sklearn.ensemble import RandomForestRegressor
forest_model = RandomForestRegressor()
forest_model.fit(X_train,y_train)

RandomForestRegressor()

In [80]:
y2_pred = forest_model.predict(X_test)

In [81]:
print('MAE for testing set: {}'.format(mean_absolute_error(y2_pred,y_test)))
print('MSE for testing set: {}'.format(mean_squared_error(y2_pred,y_test)))
print('RMSE for testing set: {}'.format(np.sqrt(mean_squared_error(y2_pred,y_test))))

MAE for testing set: 0.7091167597708716
MSE for testing set: 0.7689351942579878
RMSE for testing set: 0.8768894994570228


In [82]:
from sklearn.tree import DecisionTreeRegressor
dr=DecisionTreeRegressor()
dr.fit(X_train,y_train)

DecisionTreeRegressor()

In [83]:
y3_pred = dr.predict(X_test)

In [84]:
print('MAE for testing set: {}'.format(mean_absolute_error(y3_pred,y_test)))
print('MSE for testing set: {}'.format(mean_squared_error(y3_pred,y_test)))
print('RMSE for testing set: {}'.format(np.sqrt(mean_squared_error(y3_pred,y_test))))

MAE for testing set: 0.8641047175926472
MSE for testing set: 1.0672959724426918
RMSE for testing set: 1.033100175415091


Linear Regression is better than all having score of: MAE for testing set: 0.5715350770042953
MSE for testing set: 0.43088556545062007
RMSE for testing set: 0.6564187424583641